### PR TITLE
Add Docker security scanning with trivy

### PR DESCRIPTION
**Pull request by sldblog**
## What does this pull request do?

I came across [trivy](https://github.com/aquasecurity/trivy) in ThoughtWorks' technology radar:
https://www.thoughtworks.com/radar/tools?blipid=201911077

> Our teams particularly like Trivy, a vulnerability scanner for
> containers, because it's easier to set up than other tools, thanks to
> it shipping as a stand-alone binary.

Stand-alone setup sounded like a good fit for our use-case, so I decided
to give it a try.

## Any other changes that would benefit highlighting?

Currently, only critical severity problems cause a pipeline failure and
_only_ production requires the scan to pass.

## Checklist

- [ ] Provided JIRA ticket number in the title, e.g. "LGA-152: Sample title" **Not applicable.**

### CODE REVIEW

**Comment by exonian** [comment]
This looks fantastic!

non-blocking: wonder if it's possible to cache our built image in the `build` step and use it directly here, saving us having to fetch it again. May not be worth bothering, even if it's possible

**Comment by sldblog** [comment]
It depends on what our desired outcome is: do we stop the image from reaching the Docker repository if it has critical problems, or do we flag it as a problem for fixing until before production?

The former forces a security fix before the pipeline can proceed. I like that but didn't want to impose it on the team. I'm happy to go that way.

**Comment by exonian** [comment]
Well, there's that. And I agree with you, so think the gating you've got here is spot on, for now at least.

What I actually meant was just an optimisation thing to try to avoid having to download the image we'd just pushed, still in the same step as you are here, just changing where we source it from. But, when I come to write it out, it really doesn't seem that huge a gain (we'd still be pushing, so all it'd be is saving the fetch, and that's probably not life-changing)

**Comment by sldblog** [comment]
Oh! I misunderstood, sorry 😅 

Our current image is ~535MB. We could persist it and restore it in another job: it needs more CircleCI YAMLing and `docker save`, `docker load`.

The first job just finished and it looks like it takes 15 seconds to pull the image, which I think is good enough to not bother with the above. What do you think?

**Comment by exonian** [comment]
100% agree. Nice to have the approximate numbers, thanks.

**Comment by sldblog** [comment]
@exonian I changed a few things, could you re-review, please?

File: .circleci/config.yml [hunk 1]
```diff
@@ -95,6 +95,49 @@ jobs:
...
+      - run:
+          name: Scan for vulnerabilities (informative, non-breaking)
+          command: |
+            source .circleci/define_build_environment_variables
+            trivy --exit-code 0 --severity UNKNOWN,LOW,MEDIUM,HIGH --no-progress ${ECR_DEPLOY_IMAGE}
+      - run:
+          name: Scan for breaking vulnerabilities
+          command: |
+            source .circleci/define_build_environment_variables
+            trivy --exit-code 1 --severity CRITICAL --no-progress ${ECR_DEPLOY_IMAGE}
```
Comment (sldblog) [inline .circleci/config.yml]: At some point, we can start moving `HIGH` over here, etc.

**Comment by exonian** [comment]
Love the step splitting, great spot. All looks good to me

**Comment by sldblog** [comment]
Thank you! 🙇 🎩

**Pull request merged**